queues/c-or-cpp/double-ended-queue-using-array.c

```diff
@@ -0,0 +1,192 @@
+    }
+    // checking whether the queue has 1 data or not
+    else if(array->front==0)
+{
+ array->front=array->capacity-1;
+array->arr[array->front]=data;
+}
+// normal condition
+else
+{
+    array->front--;
+    array->arr[array->front]=data;
+}
+}
+// insertion rear basically insert data from end of a queue
+void insertion_rear(struct deque *array,int data)
+{
+    //checking is the front equal to -1 means queue is empty or not
+    if(array->rear==array->capacity-1&&array->front==0||array->rear+1==array->front)
+    printf("queue is full");
+    //checking is the rear equal to -1 means queue is empty or not
+else if(array->rear==-1)
+	{
+	    array->rear++;
+	    array->arr[array->rear]=data;
+	    array->front=array->rear;
+	}
+	// checking whether queue is full from rear but not from front
+	else if(array->rear==array->capacity-1&&array->front!=0)
+	{
+	    array->rear=0;
+	    array->arr[array->rear]=data;
+	}
+	// normal condition
+	else
+	{
+	    array->rear++;
+	    array->arr[array->rear]=data;
+	}
+
+}
+// deletion front basically delete data from front of queue
+void deletion_front(struct deque *array)
+{
+    // queue empty condition
+    if(array->front==-1)
+	printf("queue is empty");
+    else
+    {
+        // printing the value which is going to be deleted
+    printf("deleted value is %d",array->arr[array->front]);
+    // checking whether the queue has only one data
+	if(array->front==array->rear)
+	{
+	    array->front=-1;
+	    array->rear=-1;
+	}
+	// checking whether front is equal to capacity-1
+	else if(array->front==array->capacity-1)
+	    array->front=0;
+	// normal condition
+	else
+	    array->front++;
+
+
+ }
+}
+// deletion rear basically delete data from end of queue
+void deletion_rear(struct deque *array)
+{
+    //empty condition
+    if(array->rear==-1)
+      printf("queue is empty");
+      else
+      { printf("deleted value is %d",array->arr[array->rear]);
+    // checking whether the queue has only one data
+          if(array->rear==array->front)
+          {
+        array->front=-1;
+	    array->rear=-1;
+          }
+          // checking whether rear is 0 but there not only one data left that means that data are at end of queue
+          else if(array->rear==0)
+            array->rear=array->capacity-1;
+            //normal condition
+          else
+            array->rear--;
+      }
+}
+// menu basically provide the no. of task can be performed so you can choose one of them
+int menu(void)
+{
+    int choice;
+    printf("\n1: Insertion at Front");
+    printf("\n2: Insertion at Rear");
+    printf("\n3: Deletion at Front");
+    printf("\n4: Deletion at Rear");
+    printf("\n5: exit");
+    printf("\n\n enter your choice");
+    scanf("%d",&choice);
+    return(choice);
+}
+// driver code
+int main()
+{
+    struct deque *array;
+    int cap,data;
+    printf("enter the capacity");
+    scanf("%d",&cap);
+    array=create_dequeue(cap);
+    while(1)
+    { system("cls");
```

Comment:
`cls` command is windows and DOS specific, but as currently there is no standard function available in C to clear the screen so least we can do is to change the command with the different platforms. 

you could create and use a function like following to check wether the program is running on windows or not.

```c
// clear function for different platforms ( currently only windows and unix based os are supported )
void clear_screen()
{
    #ifdef _WIN32
        system("cls");
    #elif __unix__
        system("clear");
    #else
        printf("clear screen not supported\n");
    #endif
}
```

Reply:
ok i understand